content/post/mnist-example.md

```diff
@@ -0,0 +1,148 @@
+
+## Creating the models
+
+Even though we have said that the focus will be on inference, we still need to create some models to be able to use them. In this case, all the code for training is contained in `train_models.py`. We invite you to take a detailed look at it, but for the sake of this blog entry we only need to know that it is training a PCA for data preprocessing and a logistic regression for classification. The two models are then serialized to disk using sklearn's built-in pickle compatibility, to two separate file: `pca.model` and `logistic.model`.
+
+## Application Setup
+
+Our application will contain two separate computations, one for the PCA transformation and one for the classification itself.
+We set up our Wallaroo application as follows
+
+```python
+def application_setup(args):
+    with open('pca.model', 'r') as f:
+        pca = pickle.load(f)
+    with open('logistic.model', 'r') as f:
+        logistic = pickle.load(f)
+    global pca
+    global logistic
+
+    in_host, in_port = wallaroo.tcp_parse_input_addrs(args)[0]
+    out_host, out_port = wallaroo.tcp_parse_output_addrs(args)[0]
+
+    ab = wallaroo.ApplicationBuilder("MNIST classification")
+    ab.new_pipeline("MNIST",
+                    wallaroo.TCPSourceConfig(in_host, in_port, decode))
+    ab.to(pca_transform)
+    ab.to(logistic_classify)
+    ab.to_sink(wallaroo.TCPSinkConfig(out_host, out_port, encode))
+    return ab.build()
+```
+
+Note how we are loading our model in `application_setup` and making them global.  This is because loading a pickled scikit-learn model could be a potentially expensive operation, and we want to ensure we do it only once per worker. By doing so in the application_setup, the models are only loaded during initialization, and made available to the computations via the `global` keyword.
+
+### Decoder
+
+We will be sending pickled images from a specialised sender application, and we will have to unpickle them in the decoder as they are received
+
+```python
+@wallaroo.decoder(header_length=4, length_fmt=">I")
+def decode(bs):
+    return pickle.loads(bs)
+```
+
+### PCA Computation
+
+The first computation applies the PCA that has been previously loaded.
+
+```python
+@wallaroo.computation(name="PCA")
+def pca_transform(x):
+    return pca.transform([x])
+```
+
+### Logistic Regression computation
+
+The second computation applies the logistic regression, also previously loaded.
+
+```python
+@wallaroo.computation(name="Logistic Regression")
+def logistic_classify(x):
+    return logistic.predict(x)
+```
+
+### Encoder
+
+The encoder simply packs the result into a string and sends it across the wire.
```

Comment:
it doesn't though. its using a framed protocol.

either the comment or the code should be updated accordingly.